src/solvers/refinement/string_refinement.cpp

```diff
@@ -336,15 +340,37 @@ static bool has_char_pointer_subtype(const typet &type, const namespacet &ns)
   return false;
 }
 
-/// \param expr: an expression
+/// \param type: a type
 /// \param ns: name space
-/// \return true if a subexpression of `expr` is an array of characters
-static bool has_char_array_subexpr(const exprt &expr, const namespacet &ns)
+/// \return true if a subtype of `type` is string_typet.
+///         The meaning of "subtype" is in the algebraic datatype sense:
+///         for example, the subtypes of a struct are the types of its
+///         components, the subtype of a pointer is the type it points to,
+///         etc...
+static bool has_string_subtype(const typet &type, const namespacet &ns)
 {
-  for(auto it = expr.depth_begin(); it != expr.depth_end(); ++it)
-    if(is_char_array_type(it->type(), ns))
-      return true;
-  return false;
+  if(type == string_typet())
+    return true;
+
+  if(type.id() == ID_struct || type.id() == ID_union)
```

Comment:
This comment got erased but I think still applies:

Can you pull the common code between this method and `has_char_pointer_subtype` into a separate method (`has_type_have_subtype(const typet & type, const namespacet &ns, const typet &required_subtype)`)